Fix size and postion of STC popups with DirectWrite

NPDevelopmentCode opened the pull request:
As discussed in PR #1904, this is a 2 fixes for the size and position of the STC popup.  I also tried to pull cccea8f from that PR.

First, the popup was positioned incorrectly when images were used.  The text in the popup is supposed to be immediately below the text in the editor, but instead the images are below the text:

![incorrectwithimages](https://user-images.githubusercontent.com/10080894/85631319-ebe17080-b63a-11ea-8b84-228cbf368eec.png)

This was due to me not realizing that Scintilla positions the popup before adding the items to it.  To fix the issue, the images are tracked as they are registered for use with the popup, and the information is used to help Scintilla position the popup.

-----

Second, when D2D was used to do the drawing, the font in the popup could be too large:

![incorrectd2d](https://user-images.githubusercontent.com/10080894/85631503-45e23600-b63b-11ea-9897-599d8df4d030.png)

The effect was comically bad when the screen scaling was set higher than 100%.  At 200% it looks like this:

![incorrectd2d200](https://user-images.githubusercontent.com/10080894/85631586-6f02c680-b63b-11ea-87a2-5b692bc39e88.png)

This was due a confusion because font sizes with D2D are specified in DIPs instead of points.  To fix the issue, I used the same functions for drawing the popup that are used for drawing the editor instead of wxDC functions.

MaartenBent committed:
Fix the DPI in wxSTC IME window

This same modification was made to the Scintilla/SciTE source.

Commit:
Fix location of stc autocomp popup with images

The stc autocomp popup is supposed to report the location at which text
is drawn with the wxSTCListBox::CaretFromEdge function. To return the
correct value, the popup needs to compute the largest width of any
bitmap used for icons in the popup since text is drawn to the right of
the popup's images.

Previously the wxSTCListBox tried to compute this largest width when
the listbox items were set. However, Scintilla actually calls
CaretFromEdge before setting the listbox items. Consequently the
CaretFromEdge was returning a value that assumed zero width for the
list's images since images had not been set yet.

Instead, keep a running tally of the widest image width when images are
registered. This means that this variable must be moved to the
wxSTCListBoxVisualData class since that is where image registration is
handled. Also track the largest image height since that is also needed.

Commit:
Fix size of stc autocomp popup with Direct2D

Due to a confusion between DIPs and Pts, the stc autocomp window was
a little larger than it should have been when Direct2D was used.
Instead use the methods defined in SurfaceD2D so the popup will be
measured and drawn the same way the main editor window is.

MaartenBent reviewed (commented):
As I mentioned, a few small comments.

Thanks for (correctly) fixing this so quickly.

NewPagodi commented on src/stc/PlatWX.cpp:
```diff
@@ -2738,9 +2798,36 @@ void wxSTCListBox::SetContainerBorderSize(int s)
-    SetFont(*((wxFont*)font.GetID()));
-    int w;
-    GetTextExtent(EXTENT_TEST, &w, &m_textHeight);
+    if ( m_technology == wxSTC_TECHNOLOGY_DEFAULT )
```
I tried to incorporate all these comments in  6d6fbe4

MaartenBent commented on src/stc/PlatWX.cpp:
```diff
@@ -2414,10 +2417,47 @@ void wxSTCListBoxVisualData::RegisterImage(int type, const wxBitmap& bmp)
+    }
+    else
+    {
+        if ( bmp.GetWidth() > m_imgAreaSize.GetWidth() )
+        {
+            m_imgAreaSize.SetWidth(bmp.GetWidth());
+        }
+
+        if ( bmp.GetHeight() > m_imgAreaSize.GetHeight() )
+        {
+            m_imgAreaSize.SetHeight(bmp.GetHeight());
+        }
```
Use build-in `wxSize` function:
`m_imgAreaSize.IncTo(bmp.GetSize());`

NewPagodi replied:
I made these changes.  I also realized it's only necessary to recheck the entire list of images if an item is being replaced AND the new item has a different size.

MaartenBent commented:
One last issue I noticed in Travis, after this it should be all fine.
```bash
./src/stc/PlatWX.cpp:2722:11: warning: private field 'm_surfaceFontData' is not used [-Wunused-private-field]
    void* m_surfaceFontData;
          ^
1 warning generated.
```

NewPagodi commented:
I was trying to keep the `#ifdef`s to a minimum, but I guess I might as well add 2 more to completely so that variable will be completely buried.  I guess on the positive side that let's me use the proper datatype for `m_surfaceFontData` instead of `void*`.

Commit:
Add changes requested in comments for PR 1906

NewPagodi commented:
I made 1 more change.  While looking over every thing again, I noticed I was making an unnecessary copy in the SetListBoxFont method.

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -3026,24 +3075,85 @@ void wxSTCListBox::OnDrawItem(wxDC& dc, const wxRect& rect, size_t n) const
+
+        int maxWidth = rect.GetWidth() - leftGap;
+
+        if ( curWidth > maxWidth )
```
This is also pretty minor, but it's a bit confusing to see comparisons in both directions. Could we write this as a loop to avoid this? I.e. something like
```cpp
for ( int i = label.length(); curWidth > maxWidth && i; --i )
{
    ...
}
```
?

NewPagodi replied:
I made both of these changes below.

vadz reviewed (approved):
Thanks a lot for fixing this and the detailed explanations and screenshots!

I'm perfectly fine with merging this as is, but if you think it could be worth tweaking the code as indicated below, please feel free to do it (or not).

MaartenBent committed:
Move wxSTC DirectDraw specific code to derived class

MaartenBent commented:
> I was trying to keep the `#ifdef`s to a minimum, but I guess I might as well add 2 more to completely so that variable will be completely buried. I guess on the positive side that let's me use the proper datatype for `m_surfaceFontData` instead of `void*`.

What about creating a `wxSTCListBoxD2D` class that derives from `wxSTCListBox`, something [like this](https://github.com/NewPagodi/wxWidgets/compare/stctechfix2...MaartenBent:stctechfix2)? Only thing I don't like is some code duplication in `OnDrawItem`.

Commit:
Remove duplicate code in wxSTCListBox::OnDrawItem

To remove code duplication in the OnDrawItem in wxSTCListBox and
wxSTCListBoxD2D, move the code for drawing just the text to a new
virtual method named OnDrawItemText. Then have the OnDrawItem method
call the new function.

Commit:
Add other changes requested in comments for PR 1906

Commit:
Remove double strlen call in wxSTCListBoxD2D

The length of a string in wxSTCListBoxD2D::OnDrawItemText was being
calculated twice - first when measured and again when drawn. Instead
store the length the first time it is calculated. Also, pass a copy of
the string in wx2stclen to allow the calculation to work in ASCII
builds.

NewPagodi commented:
I pulled the branch creating the `wxSTCListBoxD2D` class and removed the code duplication by adding a new virtual `OnDrawItemText` function and leaving the duplicate code in `OnDrawItem`.  I also realized I was calculating the string length for the label's both when measuring and drawing the label, so I fixed that too.

MaartenBent commented on src/stc/PlatWX.cpp:
```diff
@@ -3118,10 +3220,19 @@ class wxSTCListBoxWin : public wxSTCPopupWindow
                 :wxSTCPopupWindow(parent)
 {
-    *lb = new wxSTCListBox(this, v, h);
+#ifdef HAVE_DIRECTWRITE_TECHNOLOGY
+    if ( tech == wxSTC_TECHNOLOGY_DIRECTWRITE )
+    {
+        *lb = new wxSTCListBoxD2D(this, v, h);
+    }
+    else
+#endif
+    {
+        *lb = new wxSTCListBox(this, v, h);
+    }
```
This adds a new warning:
```bash
../../src/stc/PlatWX.cpp:3223:72: warning: unused parameter 'tech' [-Wunused-parameter]
                                  wxSTCListBoxVisualData* v, int h, int tech)
                                                                        ^
```

Maybe a switch statement would be better here:
```c++
    switch ( tech )
    {
#ifdef HAVE_DIRECTWRITE_TECHNOLOGY
        case wxSTC_TECHNOLOGY_DIRECTWRITE:
            *lb = new wxSTCListBoxD2D(this, v, h);
            break;
#endif
        case wxSTC_TECHNOLOGY_DEFAULT:
        default:
            *lb = new wxSTCListBox(this, v, h);
    }
```

Commit:
Prevent warning in wxSTCListBoxWin ctr

Currently 1 argument for the constructor is only used in an #ifdef
block, and so some compilers give an unused parameter warning. Instead
use the parameter with a switch statement to avoid the warning.

Commit:
Use ellipsis char not 3 periods to shorten label

Commit:
Add comments for wxSTCListBoxD2D and its methods

NewPagodi commented:
I changed to switch statement to get rid of that warning.  

I changed the ellipsis indicator in Unicode builds from 3 periods to the horizontal ellipsis character (U+2026).  That will allow up to 2 more characters to be shown.

Finally I realized the new code is completely uncommented.  Since this is likely to be foreign even to other wxWidgets developers, I explained the main steps taken in each method.

vadz commented:
Thanks again for fixing and improving this!

Merged with 2 couple of really trivial changes that I couldn't resist making while rereading the code for one last time.
